### PR TITLE
Ledger validator (WIP)

### PR DESCRIPTION
**Pull request by thibault-martinez**
This PR implements the Ledger Validator logic.
There are still some TODO I'm working on.

### CODE REVIEW

File: ciri/core.c [hunk 1]
```diff
@@ -60,6 +60,13 @@ retcode_t core_init(core_t* const core) {
     return RC_CORE_FAILED_NODE_INIT;
   }
 
+  log_info(CORE_LOGGER_ID, "Initializing ledger validator\n");
```
Comment (tsvisabo) [inline ciri/core.c:69]: we should have one init function to init all consensus modules, next PR?

File: consensus/ledger_validator/ledger_validator.c [hunk 2]
```diff
@@ -70,3 +84,71 @@ retcode_t iota_consensus_ledger_validator_destroy(
...
+      log_error(LEDGER_VALIDATOR_LOGGER_ID, "Getting latest diff failed\n");
+      goto done;
+    }
+    if (diff == NULL) {
+      *has_snapshot = false;
+      goto done;
+    }
+    if ((ret = iota_snapshot_create_patch(
+             lv->milestone_tracker->latest_snapshot, &diff, &patch)) != RC_OK) {
+      log_error(LEDGER_VALIDATOR_LOGGER_ID, "Creating patch failed\n");
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c:305]: should we not also test for `patch != NULL` or `RC_OK` implies that always?

File: consensus/ledger_validator/ledger_validator.c [hunk 3]
```diff
@@ -17,14 +19,100 @@
 
 #define LEDGER_VALIDATOR_LOGGER_ID "consensus_ledger_validator"
 
+// TODO move ?
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c]: Yes, these are quite generic, we should have `uthash_structs.h`

File: consensus/ledger_validator/ledger_validator.c [hunk 4]
```diff
@@ -17,14 +19,100 @@
...
+typedef struct hash_queue_s {
+  flex_trit_t hash[FLEX_TRIT_SIZE_243];
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c]: We should also have MACRO/function wrappers for `poll/peek` and `add`?

File: consensus/ledger_validator/ledger_validator.c [hunk 5]
```diff
@@ -17,14 +19,100 @@
...
+  // Add milestone hash to the queue as entry point for the traversal
+  if ((hash_queue_elem = malloc(sizeof(hash_queue_t))) == NULL) {
+    ret = RC_LEDGER_VALIDATOR_OOM;
+    goto done;
+  }
+  memcpy(hash_queue_elem->hash, milestone_hash, FLEX_TRIT_SIZE_243);
+  CDL_APPEND(non_analyzed_hashes, hash_queue_elem);
+
+  while (non_analyzed_hashes != NULL) {
+    tx_hash.trits = non_analyzed_hashes->hash;
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c]: `peek`

File: consensus/ledger_validator/ledger_validator.c [hunk 6]
```diff
@@ -17,14 +19,100 @@
...
+      hash_pack_reset(&pack);
+      if ((ret = iota_tangle_transaction_load(lv->tangle, TRANSACTION_COL_HASH,
+                                              &tx_hash, &pack)) != RC_OK) {
+        goto done;
+      }
+      if (tx.snapshot_index == 0) {
+        // TODO update snapshot index
+        // TODO messageQ publish
+        // Add trunk hash to the queue
+        if ((hash_queue_elem = malloc(sizeof(hash_queue_t))) == NULL) {
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c]: `add` (should do the malloc/memcpy/CDL_APPEND)

File: consensus/ledger_validator/ledger_validator.c [hunk 7]
```diff
@@ -17,14 +19,100 @@
...
+    tmp = non_analyzed_hashes;
+    CDL_DELETE(non_analyzed_hashes, non_analyzed_hashes);
+    free(tmp);
+  }
+
+done : {
+  hash_queue_t *iter_q, *tmp_q1, *tmp_q2;
+  hash_set_t *iter_s, *tmp_s;
+
+  CDL_FOREACH_SAFE(non_analyzed_hashes, iter_q, tmp_q1, tmp_q2) {
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c]: let's have destory methods for uthash set+queue

File: consensus/ledger_validator/ledger_validator.c [hunk 8]
```diff
@@ -121,29 +123,50 @@ done : {
...
+    //                 stateDiffViewModel.getDiff()))) {
+    //           milestoneTracker.latestSnapshot.apply(stateDiffViewModel.getDiff(),
+    //                                                 candidateMilestone.index());
+    *consistent_index = milestone.index;
+    memcpy(consistent_hash, milestone.hash, FLEX_TRIT_SIZE_243);
+    //         } else {
+    //           break;
+    //         }
+    //       }
+    hash_pack_reset(&pack);
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c:114]: Can we load all milestones with one db access before looping?

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: consensus/ledger_validator/ledger_validator.c [hunk 9]
```diff
@@ -167,6 +167,138 @@ static retcode_t build_snapshot(ledger_validator_t *const lv,
...
+        if (tx.current_index == 0) {
+          valid_bundle = false;
+          if ((ret = bundle_validate(lv->tangle, &tx_hash, bundle,
+                                     &valid_bundle)) != RC_OK) {
+            goto done;
+          }
+          if (valid_bundle &&
+              (tx_bundle = (iota_transaction_t)utarray_eltptr(bundle, 0)) !=
+                  NULL &&
+              memcmp(tx_bundle->hash, tx_hash.trits, FLEX_TRIT_SIZE_243) == 0) {
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c]: i think we should include this validation (`memcmp`) in `bundle_validate`, @th0br0  ?

File: consensus/ledger_validator/ledger_validator.c [hunk 10]
```diff
@@ -167,6 +167,138 @@ static retcode_t build_snapshot(ledger_validator_t *const lv,
...
+              tx_bundle = (iota_transaction_t)utarray_next(bundle, tx_bundle);
+            }
+          }
+          if (!valid_bundle) {
+            *valid_diff = false;
+            goto done;
+          }
+        }
+        // Add trunk hash to the queue
+        if ((hash_queue_elem = malloc(sizeof(hash_queue_t))) == NULL) {
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c]: `uthash_queue_add`

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: ciri/core.c (hunk 1 above)
Comment (thibault-martinez) [reply ciri/core.c:69]: Yes, I've been willing to do that for a while, but yes another PR to keep this clear.

File: consensus/ledger_validator/ledger_validator.c (hunk 3 above)
Comment (thibault-martinez) [reply consensus/ledger_validator/ledger_validator.c]: In the most recent commit, they have been moved to "/utils/traversal.h", where we'll put traversal utilities, is that ok ?

File: consensus/ledger_validator/ledger_validator.c (hunk 4 above)
Comment (thibault-martinez) [reply consensus/ledger_validator/ledger_validator.c]: Sure, I'm looking into it rn

File: consensus/ledger_validator/ledger_validator.c (hunk 2 above)
Comment (thibault-martinez) [reply consensus/ledger_validator/ledger_validator.c:305]: Yes, we should

Comment (thibault-martinez) [reply consensus/ledger_validator/ledger_validator.c:305]: After thoughts, it can create a `NULL` patch, and that's not a problem, no need to check

File: consensus/ledger_validator/ledger_validator.c [hunk 11]
```diff
@@ -150,13 +149,12 @@ static retcode_t get_latest_diff(ledger_validator_t *const lv,
     goto done;
   }
 
-  // Add null hash to the set of visited hashes
-  if ((hash_set_elem = malloc(sizeof(hash_set_t))) == NULL) {
-    ret = RC_LEDGER_VALIDATOR_OOM;
-    goto done;
+  {
+    flex_trit_t null_hash[FLEX_TRIT_SIZE_243] = {FLEX_TRIT_NULL_VALUE};
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c]: does that initialize the entire array or just sets the first element?

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: consensus/ledger_validator/ledger_validator.c [hunk 12]
```diff
@@ -177,10 +177,8 @@ static retcode_t get_latest_diff(ledger_validator_t *const lv,
                                      &valid_bundle)) != RC_OK) {
             goto done;
           }
-          if (valid_bundle &&
-              (tx_bundle = (iota_transaction_t)utarray_eltptr(bundle, 0)) !=
-                  NULL &&
-              memcmp(tx_bundle->hash, tx_hash.trits, FLEX_TRIT_SIZE_243) == 0) {
+          if (valid_bundle && (tx_bundle = (iota_transaction_t)utarray_eltptr(
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c:180]: call it tail?

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: consensus/ledger_validator/ledger_validator.c (hunk 12 above)
Comment (thibault-martinez) [reply consensus/ledger_validator/ledger_validator.c:180]: It's not always the tail, we are looping through the bundle with it

File: consensus/ledger_validator/ledger_validator.c [hunk 13]
```diff
@@ -333,3 +333,86 @@ retcode_t iota_consensus_ledger_validator_update_snapshot(
...
+                                        &current_state, &patch)) != RC_OK) {
+    log_error(LEDGER_VALIDATOR_LOGGER_ID, "Creating patch failed\n");
+    goto done;
+  }
+  *is_consistent = iota_snapshot_is_state_consistent(&patch);
+
+  if (is_consistent) {
+    // TODO Extract state_map_t from snapshot #345
+    HASH_ITER(hh, current_state, iter, tmp) {
+      HASH_FIND(hh, *diff, iter->hash, FLEX_TRIT_SIZE_243, entry);
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c:418]: There's a little code duplication here, can we have a static function  that does that?

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: consensus/ledger_validator/ledger_validator.c (hunk 13 above)
Comment (thibault-martinez) [reply consensus/ledger_validator/ledger_validator.c:418]: I'd prefer to deal with that in a separate PR since it needs to move some code away from snapshot (https://github.com/iotaledger/entangled/issues/345)
And the two loops are slightly different

File: consensus/ledger_validator/ledger_validator.c (hunk 11 above)
Comment (thibault-martinez) [reply consensus/ledger_validator/ledger_validator.c]: Whoops

File: consensus/ledger_validator/ledger_validator.c (hunk 13 above)
Comment (tsvisabo) [reply consensus/ledger_validator/ledger_validator.c:418]: Alrighty

File: consensus/ledger_validator/ledger_validator.c (hunk 12 above)
Comment (tsvisabo) [reply consensus/ledger_validator/ledger_validator.c:180]: Fair enough

File: consensus/ledger_validator/ledger_validator.c [hunk 14]
```diff
@@ -333,6 +332,25 @@ retcode_t iota_consensus_ledger_validator_update_snapshot(
   }
   return ret;
 }
+
+retcode_t iota_consensus_ledger_validator_check_consistency(
+    ledger_validator_t *const lv, hash_list_t **hashes, bool *consistent) {
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c]: I dont think we need `hashes` as `**`

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: consensus/ledger_validator/ledger_validator.c [hunk 15]
```diff
@@ -333,3 +333,86 @@ retcode_t iota_consensus_ledger_validator_update_snapshot(
   }
   return ret;
 }
+retcode_t iota_consensus_ledger_validator_update_diff(
+    ledger_validator_t *const lv, hash_set_t **analyzed_hashes,
+    state_map_t *diff, flex_trit_t *tip, bool *is_consistent) {
+  retcode_t ret = RC_OK;
+  state_map_t current_state = NULL, patch = NULL;
+  state_entry_t *iter, *entry, *new, *tmp;
```
Comment (tsvisabo) [inline consensus/ledger_validator/ledger_validator.c]: Let's not call it `new` :)

**Review by tsvisabo — CHANGES_REQUESTED** [review]

**Review by tsvisabo — APPROVED** [review]
